src/moaicore/MOAICCParticleSystem.cpp

```diff
@@ -0,0 +1,2384 @@
+    
+    self->InitializeEmitter();
+    
+    return 0;
+}
+
+//----------------------------------------------------------------//
+/** @name   load
+    @text   Initialize particle system properties with an XML file.
+ 
+    @in     MOAICCParticleSystem    self
+    @in     String                  file to load
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_load( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "US" )
+    cc8* xml = lua_tostring( state, 2);
+    
+    if ( MOAILogMessages::CheckFileExists ( xml, L )) {
+        TiXmlDocument doc;
+        doc.LoadFile ( xml );
+        self->ParseXML ( xml, doc.RootElement ());
+        
+    }
+    
+    return 0;
+}
+//----------------------------------------------------------------//
+/** @name   startSystem
+    @text   Begin the generation of particles.  Use in conjenction with start().
+ 
+    @in     MOAICCParticleSystem    self
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_startSystem ( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "U" )
+    self->StartSystem();
+    return 0;
+}
+
+//----------------------------------------------------------------//
+/** @name   stopSystem
+    @text   Halt the generation of particles.  Any remaining particles will be allowed to stay until the end of their lifespan. Use stop() to halt the animation of particles.
+ 
+    @in     MOAICCParticleSystem    self
+    @out    nil
+ */
+int MOAICCParticleSystem::_stopSystem ( lua_State *L ) {
+    
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "U" )
+    self->StopSystem();
+    
+    return 0;
+}
+//----------------------------------------------------------------//
+/** @name   surge
+    @text   Generate the given number of particles at one time.
+    @in     MOAICCParticleSystem    self
+    @in     number                  numberOfParticles
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_surge ( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "UN" )
+    u32 numberOfParticles = state.GetValue < u32 >(2, 1.0);
+    self->Surge( numberOfParticles );
+    return 0;
+}
+//----------------------------------------------------------------//
+/** @name   reset
+    @text   Make all particles get cleared out at the next frame.
+ 
+    @in     MOAICCParticleSystem    self
+    @opt    boolean                 activate    Default true.  Whether to begin the generation of particles after clearing the particles.
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_reset ( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "U" )
+    bool activate = state.GetValue < bool > (2, true);
+    self->ResetSystem( activate );
+    return 0;
+}
+
+//----------------------------------------------------------------//
+//----------------------------------------------------------------//
+
+bool MOAICCParticleSystem::AddParticle () {
+    if ( this->IsFull() ) {
+        return false;
+    }
+    
+    MOAICCParticle particle(this);
+    this->mParticles.push_back(particle);
```

Comment:
This will invoke the copy constructor because `particle` is an lvalue. If you do this:

    this->mParticles.push_back(MOAICCParticle(this));

It will now look for a move constructor (provided MOAICCParticle has a move constructor).

More on move semantics: http://www.cprogramming.com/c++11/rvalue-references-and-move-semantics-in-c++11.html